### DISCUSSION TITLE
onAuthStateChange can't capture PASSWORD_RECOVERY events? (Using React)

### DISCUSSION THREAD

**Discussion by luketpena** [supabase/supabase#12229] [Questions]
In the documentation example for setting up Supabase auth in a React project, it shows us setting up the listener for this event in a useEffect in the App.tsx. We have done this in our project.

However, these events do not fire on app init, only on subsequent auth event changes. For example, if I visit the page and I am already authenticated, I will not see a SIGNED_IN event. But if I navigate in my OS away from my browser and come back, it will fire the event.

This becomes a problem with the PASSWORD_RECOVERY event. When we click the "reset password" link in the email, it navigates us to the app, authenticates us, and then (according to the docs) should trigger two auth events: SIGNED_IN and PASSWORD_RECOVERY.

These events _must_ fire on app init, but no events are triggered on init. Or at least that is the behavior we are observing and is corroborated by [this StackOverflow post.](https://stackoverflow.com/questions/72385641/supabase-onauthstatechanged-how-do-i-properly-wait-for-the-request-to-finish-p)

Currently, we are considering sending a redirectTo URL with a unique hash that we track in our database related to a particular user and a reset request, but that seems like a crazy work around to a problem that has a solution right there in the docs. It just isn't working. What are we missing? Why can we not capture these events properly? Happy to provide any additional information that might help.

**Comment by thorwebdev** [comment]
The `SIGNED_IN` event is only triggered when there's tokens in the fragment (`detectSessionInUrl`), e.g. after a Oauth / magic link / password recovery link etc. So when initialising supabase-js in the `useEffect` it should detect the fragment in the URL and first fire the `SIGNED_IN` event and then the `PASSWORD_RECOVERY`: https://github.com/supabase/gotrue-js/blob/master/src/GoTrueClient.ts#L199-L215

There might be some additional React specific nuances here: https://github.com/supabase/gotrue-js/issues/518#issuecomment-1367971483 

Do you happen to have your code or a minimal reproducible example somewhere?
